doc/ranger.1
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "RANGER 1"
.TH RANGER 1 "power-ranger-1.9.2" "2018-10-26" "power-ranger manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
power-ranger \- visual file manager
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBpower-ranger\fR [\fB\-\-version\fR] [\fB\-\-help\fR] [\fB\-\-debug\fR] [\fB\-\-clean\fR]
[\fB\-\-cachedir\fR=\fIdirectory\fR] [\fB\-\-confdir\fR=\fIdirectory\fR]
[\fB\-\-datadir\fR=\fIdirectory\fR] [\fB\-\-copy\-config\fR=\fIwhich\fR]
[\fB\-\-choosefile\fR=\fItarget\fR] [\fB\-\-choosefiles\fR=\fItarget\fR]
[\fB\-\-choosedir\fR=\fItarget\fR] [\fB\-\-selectfile\fR=\fIfilepath\fR]
[\fB\-\-show\-only\-dirs\fR]
[\fB\-\-list\-unused\-keys\fR] [\fB\-\-list\-tagged\-files\fR=\fItag\fR]
[\fB\-\-profile\fR] [\fB\-\-cmd\fR=\fIcommand\fR] [\fIpath ...\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
power-ranger is a console file manager with \s-1VI\s0 key bindings.
.SH "RESOURCES"
.IX Header "RESOURCES"
\&\fIThis manual\fR contains instructions on how to use and configure power-ranger.
.PP
\&\fIInside power-ranger\fR, you can press \fI?\fR for a list of key bindings, commands or
settings.
.PP
The \fI\s-1README\s0\fR contains install instructions.
.PP
The file \fI\s-1HACKING\s0.md\fR contains guidelines for code modification.
.PP
The directory \fIdoc/configs\fR contains configuration files.  They are usually
installed to \fI/usr/share/doc/power-ranger/config\fR and can be obtained with power-ranger's
\&\-\-copy\-config option.
.PP
The directory \fIexamples\fR contains reference implementations for power-ranger
plugins, sample configuration files and some programs for integrating power-ranger
with other software.  They are usually installed to
\&\fI/usr/share/doc/power-ranger/examples\fR.
.PP
The man page of \fBrifle\fR\|(1) describes the functions of the file opener
.PP
The section \fI\s-1LINKS\s0\fR of this man page contains further resources.
.SH "POSITIONAL ARGUMENTS"
.IX Header "POSITIONAL ARGUMENTS"
.IP "\fIpath ...\fR" 14
.IX Item "path ..."
Each path will be opened in a tab and if the path is a file it will be selected.
Omitting this is equivalent to providing the current directory.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-d\fR, \fB\-\-debug\fR" 14
.IX Item "-d, --debug"
Activate the debug mode: Whenever an error occurs, power-ranger will exit and print a
full traceback.  The default behavior is to merely print the name of the
exception in the statusbar/log and try to keep running.
.IP "\fB\-c\fR, \fB\-\-clean\fR" 14
.IX Item "-c, --clean"
Activate the clean mode:  power-ranger will not access or create any configuration
files nor will it leave any traces on your system.  This is useful when your
configuration is broken, when you want to avoid clutter, etc.
.IP "\fB\-\-cachedir\fR=\fIdir\fR" 14
.IX Item "--cachedir=dir"
Change the cache directory of power-ranger from \f(CW$XDG_CACHE_HOME\fR or ~/.cache/power-ranger to \*(L"dir\*(R".
.IP "\fB\-r\fR \fIdir\fR, \fB\-\-confdir\fR=\fIdir\fR" 14
.IX Item "-r dir, --confdir=dir"
Change the configuration directory of power-ranger from \f(CW$XDG_CONFIG_HOME\fR or ~/.config/power-ranger to \*(L"dir\*(R".
.IP "\fB\-\-datadir\fR=\fIdir\fR" 14
.IX Item "--datadir=dir"
Change the data directory of power-ranger from \f(CW$XDG_DATA_HOME\fR or ~/.local/share/power-ranger to \*(L"dir\*(R".
.IP "\fB\-\-copy\-config\fR=\fIfile\fR" 14
.IX Item "--copy-config=file"
Create copies of the default configuration files in your local configuration
directory.  Existing ones will not be overwritten.  Possible values: \fIall\fR,
\&\fIcommands\fR, \fIcommands_full\fR, \fIrc\fR, \fIrifle\fR, \fIscope\fR.
.Sp
Note: You may want to disable loading of the global configuration files by
exporting \fIRANGER_LOAD_DEFAULT_RC=FALSE\fR in your environment.  See also:
\&\fB\s-1FILES\s0\fR, \fB\s-1ENVIRONMENT\s0\fR
.Sp
\&\-\-copy\-config=\fBcommands\fR will copy only a small sample configuration file with
a thoroughly commented example.  It is recommended to keep this file tidy to
avoid getting defunct commands on power-ranger upgrades.  The full default
commands.py can be copied with \-\-copy\-config=\fBcommands_full\fR, but that file
will be ignored by power-ranger and serves only as a reference for making your own
commands.
.IP "\fB\-\-choosefile\fR=\fItargetfile\fR" 14
.IX Item "--choosefile=targetfile"
Allows you to pick a file with power-ranger.  This changes the behavior so that when
you open a file, power-ranger will exit and write the absolute path of that file into
\&\fItargetfile\fR.
.IP "\fB\-\-choosefiles\fR=\fItargetfile\fR" 14
.IX Item "--choosefiles=targetfile"
Allows you to pick multiple files with power-ranger.  This changes the behavior so
that when you open a file, power-ranger will exit and write the absolute paths of all
selected files into \fItargetfile\fR, adding one newline after each filename.
.IP "\fB\-\-choosedir\fR=\fItargetfile\fR" 14
.IX Item "--choosedir=targetfile"
Allows you to pick a directory with power-ranger.  When you exit power-ranger, it will
write the last visited directory into \fItargetfile\fR.
.IP "\fB\-\-selectfile\fR=\fItargetfile\fR" 14
.IX Item "--selectfile=targetfile"
Open power-ranger with \fItargetfile\fR selected. This is a legacy option, superseded by
the behavior for the \s-1POSITIONAL ARGUMENTS.\s0
.IP "\fB\-\-show\-only\-dirs\fR" 14
.IX Item "--show-only-dirs"
Display only the directories. May be used in conjunction with
\&\fB\-\-choosedir\fR=\fItargetfile\fR.
.IP "\fB\-\-list\-unused\-keys\fR" 14
.IX Item "--list-unused-keys"
List common keys which are not bound to any action in the \*(L"browser\*(R" context.
This list is not complete, you can bind any key that is supported by curses:
use the key code returned by \f(CW\*(C`getch()\*(C'\fR.
.IP "\fB\-\-list\-tagged\-files\fR=\fItag\fR" 14
.IX Item "--list-tagged-files=tag"
List all files which are tagged with the given tag.  Note: Tags are single
characters.  The default tag is \*(L"*\*(R"
.IP "\fB\-\-profile\fR" 14
.IX Item "--profile"
Print statistics of \s-1CPU\s0 usage on exit.
.IP "\fB\-\-cmd\fR=\fIcommand\fR" 14
.IX Item "--cmd=command"
Execute the command after the configuration has been read.  Use this option
multiple times to run multiple commands.
.IP "\fB\-\-version\fR" 14
.IX Item "--version"
Print the version and exit.
.IP "\fB\-h\fR, \fB\-\-help\fR" 14
.IX Item "-h, --help"
Print a list of options and exit.
.SH "CONCEPTS"
.IX Header "CONCEPTS"
This part explains how certain parts of power-ranger work and how they can be used
efficiently.
.SS "\s-1TAGS\s0"
.IX Subsection "TAGS"
Tags are single characters which are displayed left of a filename.  You can use
tags however you want.  Press \*(L"t\*(R" to toggle tags and \*(L"ut\*(R" to remove any tags of
the selection. The default tag is an Asterisk (\*(L"*\*(R"), but you can use any tag by
typing \fI"<tagname>\fR.
.SS "\s-1PREVIEWS\s0"
.IX Subsection "PREVIEWS"
By default, only text files are previewed, but you can enable external preview
scripts by setting the option \f(CW\*(C`use_preview_script\*(C'\fR and \f(CW\*(C`preview_files\*(C'\fR to true.
.PP
This default script is \fI\f(CI%power-rangerdir\fI/data/scope.sh\fR. It contains more
documentation and calls to the programs \fIlynx\fR and \fIelinks\fR for html,
\&\fIhighlight\fR for text/code, \fIimg2txt\fR for images, \fIatool\fR for archives,
\&\fIpdftotext\fR or \fImutool\fR for PDFs and \fImediainfo\fR for video and audio files.
.PP
Install these programs (just the ones you need) and scope.sh will automatically
use them.
.PP
Independently of the preview script, there is a feature to preview images
by drawing them directly into the terminal. To enable this feature, set the
option \f(CW\*(C`preview_images\*(C'\fR to true and enable one of the image preview modes:
.PP
\fIw3m\fR
.IX Subsection "w3m"
.PP
This does not work over ssh, requires certain terminals (tested on \*(L"xterm\*(R" and
\&\*(L"urxvt\*(R") and is incompatible with tmux, although it works with screen.
.PP
To enable this feature, install the program \*(L"w3m\*(R" and set the option
\&\f(CW\*(C`preview_images_method\*(C'\fR to w3m.
.PP
When using a terminal with a nonzero border which is not automatically detected, the w3m preview will be misaligned.
Use the \f(CW\*(C`w3m_offset\*(C'\fR option to manually adjust the image offset. This should be the same value as the terminal's border value.
.PP
\fIiTerm2\fR
.IX Subsection "iTerm2"
.PP
This only works in iTerm2 compiled with image preview support, but works over
ssh.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to iterm2.
.PP
This feature relies on the dimensions of the terminal's font.  By default, a
width of 8 and height of 11 are used.  To use other values, set the options
\&\f(CW\*(C`iterm2_font_width\*(C'\fR and \f(CW\*(C`iterm2_font_height\*(C'\fR to the desired values.
.PP
\fIterminology\fR
.IX Subsection "terminology"
.PP
This only works in terminology. It can render vector graphics, but works only locally.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to terminology.
.PP
\fIurxvt\fR
.IX Subsection "urxvt"
.PP
This only works in urxvt compiled with pixbuf support. Does not work over ssh.
.PP
Essentially this mode sets an image as a terminal background temporarily, so it
will break any previously set image background.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to urxvt.
.PP
\fIurxvt-full\fR
.IX Subsection "urxvt-full"
.PP
The same as urxvt but utilizing not only the preview pane but the whole terminal
window.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to urxvt-full.
.PP
\fIkitty\fR
.IX Subsection "kitty"
.PP
This only works on Kitty. It requires \s-1PIL\s0 (or pillow) to work.
Allows remote image previews, for example in an ssh session.
.PP
To enable this feature, set the option \f(CW\*(C`preview_images_method\*(C'\fR to kitty.
.SS "\s-1SELECTION\s0"
.IX Subsection "SELECTION"
The \fIselection\fR is defined as \*(L"All marked files \s-1IF THERE ARE ANY,\s0 otherwise
the current file.\*(R"  Be aware of this when using the :delete command, which
deletes all files in the selection.
.PP
You can mark files by pressing <Space>, v, etc.  A yellow \fBMrk\fR symbol at the
bottom right indicates that there are marked files in this directory.
.SS "\s-1MACROS\s0"
.IX Subsection "MACROS"
Macros can be used in commands to abbreviate things.
.PP
.Vb 6
\& %f   the highlighted file
\& %d   the path of the current directory
\& %s   the selected files in the current directory
\& %t   all tagged files in the current directory
\& %c   the full paths of the currently copied/cut files
\& %p   the full paths of selected files
.Ve
.PP
The macros \f(CW%f\fR, \f(CW%d\fR, \f(CW%p\fR, and \f(CW%s\fR also have upper case variants, \f(CW%F\fR, \f(CW%D\fR, \f(CW%P\fR, and
\&\f(CW%S\fR, which refer to the next tab.  To refer to specific tabs, add a number in
between.  (%7s = selection of the seventh tab.)
.PP
\&\f(CW%c\fR is the only macro which ranges out of the current directory. So you may
\&\*(L"abuse\*(R" the copying function for other purposes, like diffing two files which
are in different directories:
.PP
.Vb 2
\& Yank the file A (type yy), move to the file B, then type
\& @diff %c %f
.Ve
.PP
Macros for file paths are generally shell-escaped so they can be used in the
\&\f(CW\*(C`shell\*(C'\fR command.
.PP
Additionally, if you create a key binding that uses <any>, a special statement
which accepts any key, then the macro \f(CW%any\fR (or \f(CW%any0\fR, \f(CW%any1\fR, \f(CW%any2\fR, ...) can be
used in the command to get the key that was pressed.
.PP
The macro \f(CW%power-rangerdir\fR expands to the directory of power-ranger's python library, you
can use it for something like this command:
  alias show_commands shell less \f(CW%power-rangerdir\fR/config/commands.py
.PP
\&\f(CW%confdir\fR expands to the directory given by \fB\-\-confdir\fR.
.PP
\&\f(CW%datadir\fR expands to the directory given by \fB\-\-datadir\fR.
.PP
The macro \f(CW%space\fR expands to a space character. You can use it to add spaces to
the end of a command when needed, while preventing editors to strip spaces off
the end of the line automatically.
.PP
To write a literal %, you need to escape it by writing %%.
.SS "\s-1BOOKMARKS\s0"
.IX Subsection "BOOKMARKS"
Type \fBm<key>\fR to bookmark the current directory. You can re-enter this
directory by typing \fB`<key>\fR. <key> can be any letter or digit.  Unlike vim,
both lowercase and uppercase bookmarks are persistent.
.PP
Each time you jump to a bookmark, the special bookmark at key ` will be set
to the last directory. So typing \*(L"``\*(R" gets you back to where you were before.
.PP
Bookmarks are selectable when tabbing in the :cd command.
.PP
Note: The bookmarks ' (Apostrophe) and ` (Backtick) are the same.
.SS "\s-1RIFLE\s0"
.IX Subsection "RIFLE"
Rifle is the file opener of power-ranger.  It can be used as a standalone program or
a python module.  It is located at \fIpower-ranger/ext/rifle.py\fR.  In contrast to
other, more simple file openers, rifle can automatically find installed
programs so it can be used effectively out of the box on a variety of systems.
.PP
It's configured in \fIrifle.conf\fR through a list of conditions and commands.
For each line the conditions are checked and if they are met, the respective
command is taken into consideration.  By default, simply the first matching
rule is used.  In power-ranger, you can list and choose rules by typing \*(L"r\*(R" or simply
by typing \*(L"<rulenumber><enter>\*(R".  If you use rifle standalone, you can list all
rules with the \*(L"\-l\*(R" option and pick a rule with \*(L"\-p <number>\*(R".
.PP
The rules, along with further documentation, are contained in
\&\fIpower-ranger/config/rifle.conf\fR.
.SS "\s-1FLAGS\s0"
.IX Subsection "FLAGS"
Flags give you a way to modify the behavior of the spawned process.  They are
used in the commands \f(CW\*(C`:open_with\*(C'\fR (key \*(L"r\*(R") and \f(CW\*(C`:shell\*(C'\fR (key \*(L"!\*(R").
.PP
.Vb 4
\& f   Fork the process.  (Run in background)
\& c   Run the current file only, instead of the selection
\& r   Run application with root privilege (requires sudo)
\& t   Run application in a new terminal window
.Ve
.PP
There are some additional flags that can currently be used only in the \f(CW\*(C`shell\*(C'\fR
command: (for example \f(CW\*(C`:shell \-w df\*(C'\fR)
.PP
.Vb 3
\& p   Redirect output to the pager
\& s   Silent mode.  Output will be discarded.
\& w   Wait for an Enter\-press when the process is done
.Ve
.PP
By default, all the flags are off unless specified otherwise in the
\&\fIrifle.conf\fR configuration file.  You can specify as many flags as you want.
An uppercase flag negates the effect: \*(L"ffcccFsf\*(R" is equivalent to \*(L"cs\*(R".
.PP
The terminal program name for the \*(L"t\*(R" flag is taken from the environment
variable \f(CW$TERMCMD\fR.  If it doesn't exist, it tries to extract it from \f(CW$TERM\fR,
uses \*(L"x\-terminal-emulator\*(R" as a fallback, and then \*(L"xterm\*(R" if that fails.
.PP
Examples: \f(CW\*(C`:open_with c\*(C'\fR will open the file that you currently point at, even
if you have selected other files.  \f(CW\*(C`:shell \-w df\*(C'\fR will run \*(L"df\*(R" and wait for
you to press Enter before switching back to power-ranger.
.SS "\s-1PLUGINS\s0"
.IX Subsection "PLUGINS"
power-ranger's plugin system consists of python files which are located in
\&\fI~/.config/power-ranger/plugins/\fR and are imported in alphabetical order when
starting power-ranger.  A plugin changes power-rangers behavior by overwriting or extending
a function that power-ranger uses.  This allows you to change pretty much every part
of power-ranger, but there is no guarantee that things will continue to work in
future versions as the source code evolves.
.PP
Adding new commands via a plugin as simple as specifying them like you would do
in the \fIcommands.py\fR.
.PP
There are some hooks that are specifically made for the use in plugins.  They
are functions that start with hook_ and can be found throughout the code.
.PP
.Vb 1
\& grep \*(Aqdef hook_\*(Aq \-r /path/to/power-rangers/source
.Ve
.PP
Also try:
.PP
.Vb 1
\& pydoc power-ranger.api
.Ve
.PP
Note that you should \s-1NOT\s0 simply overwrite a function unless you know what
you're doing.  Instead, save the existing function and call it from your new
one.  This way, multiple plugins can use the same hook.  There are several
sample plugins in the \fI/usr/share/doc/power-ranger/examples/\fR directory, including a
hello-world plugin that describes this procedure.
.SH "KEY BINDINGS"
.IX Header "KEY BINDINGS"
Key bindings are defined in the file \fI\f(CI%power-rangerdir\fI/config/rc.conf\fR.  Check this
file for a list of all key bindings.  You can copy it to your local
configuration directory with the \-\-copy\-config=rc option.
.PP
Many key bindings take an additional numeric argument.  Type \fI5j\fR to move
down 5 lines, \fI2l\fR to open a file in mode 2, \fI10<Space>\fR to mark 10 files.
.PP
This list contains the most useful bindings:
.SS "\s-1MAIN BINDINGS\s0"
.IX Subsection "MAIN BINDINGS"
.IP "h, j, k, l" 14
.IX Item "h, j, k, l"
Move left, down, up or right
.IP "^D or J, ^U or K" 14
.IX Item "^D or J, ^U or K"
Move a half page down, up
.IP "H, L" 14
.IX Item "H, L"
Move back and forward in the history
.IP "gg" 14
.IX Item "gg"
Move to the top
.IP "G" 14
.IX Item "G"
Move to the bottom
.IP "[, ]" 14
Move up and down in the parent directory.
.IP "^R" 14
.IX Item "^R"
Reload everything
.IP "F" 14
.IX Item "F"
Toggle \fIfreeze_files\fR setting.  When active (indicated by a cyan \fI\s-1FROZEN\s0\fR
message in the status bar), directories and files will not be loaded, improving
performance when all the files you need are already loaded.  This does not
affect file previews, which can be toggled with \fIzI\fR.  Also try disabling the
preview of directories with \fIzP\fR.
.IP "^L" 14
.IX Item "^L"
Redraw the screen
.IP "i" 14
.IX Item "i"
Inspect the current file in a bigger window.
.IP "E" 14
.IX Item "E"
Edit the current file in \f(CW$VISUAL\fR otherwise \f(CW$EDITOR\fR otherwise \*(L"vim\*(R"
.IP "S" 14
.IX Item "S"
Open a shell in the current directory
.IP "?" 14
Opens this man page
.IP "W" 14
.IX Item "W"
Opens the log window where you can review messages that pop up at the bottom.
.IP "w" 14
.IX Item "w"
Opens the task window where you can view and modify background processes that
currently run in power-ranger.  In there, you can type \*(L"dd\*(R" to abort a process and
\&\*(L"J\*(R" or \*(L"K\*(R" to change the priority of a process.  Only one process is run at a
time.
.IP "^C" 14
.IX Item "^C"
Stop the currently running background process that power-ranger has started, like
copying files, loading directories or file previews.
.IP "<octal>=, +<who><what>, \-<who><what>" 14
.IX Item "<octal>=, +<who><what>, -<who><what>"
Change the permissions of the selection.  For example, \f(CW\*(C`777=\*(C'\fR is equivalent to
\&\f(CW\*(C`chmod 777 %s\*(C'\fR, \f(CW\*(C`+ar\*(C'\fR does \f(CW\*(C`chmod a+r %s\*(C'\fR, \f(CW\*(C`\-ow\*(C'\fR does \f(CW\*(C`chmod o\-w %s\*(C'\fR etc.
.IP "yy" 14
.IX Item "yy"
Copy (yank) the selection, like pressing Ctrl+C in modern \s-1GUI\s0 programs.  (You
can also type \*(L"ya\*(R" to add files to the copy buffer, \*(L"yr\*(R" to remove files again,

CHANGELOG.md
This log documents changes between stable versions.

# 2018-09-09: version 1.9.2
* Added a `hint_collapse_threshold` setting
* Added a `traverse_backwards` command analogous to `traverse`
* Added a command to shift tabs
* Added a normal mode mapping to quickly enter the console and scroll through
  the history `C-p`
* Added a section to `scope.sh` for image previews of archives
* Added an avfs plugin
* Added an option to the move command to enable launching the selected file
  instead of the marked files
* Added filtering functionality inspired by dired's filter stack, `.n, .| ...`
* Added image preview method for Kitty
* Added option to disable the display of free space for high latency situations
* Added section to `scope.sh` for pdf previews with mutool
* Added several emacs/readline-inspired keybindings, `C-g` for `ESC`, `alt-f/b`
* Added systemwide `rc.conf` and `commands.py` in `/etc/power-ranger`
* Added the `%any_path` macro to allow bookmarks to be used with commands that
  need a path and are unaware of bookmarks
* Added versioning logic to include extra info in unreleased versions
* Change tab saving to save all tabs, not just the active tab
* Changed `draw_borders` setting to enable drawing only borders or seperators
* Changed behavior of positional arguments to the power-ranger command, if you
  specify a path to a file power-ranger will open with that file selected
* Changed the `tilde_in_titlebar` setting to influence the window titlebar too
* Changed the default colorscheme to work properly in terminals that don't
  equate bold and bright
* Fixed StopIteration errors
* Fixed embedded null errors
* Fixed issues reported by coverity scan
* Fixed running power-ranger as root on Mac OS
* Fixed unicode issue for python2
* Fixed w3m preview issues with black stripes
* Improved PEP8 adherence
* Improved VCS symbols
* Improved `--cmd` functionality
* Improved file encoding detection by using chardet if it's available
* Rifle's flag t should now work with more terminals than xterm and urxvt
* Update colorscheme documentation

# 2018-02-22: version 1.9.1
* Fixed the rifle config backwards compatibility (regression in 1.9.0)
* Fixed the POSIX compatibility of `Makefile`
* Fixed `--choosefile`, `--choosefiles` and `--choosedir` so they work
  with the process substitution (`>(...)` in Bash)
* Changed the default `gt` binding to `gp` due to a conflict
* Changed the help message for `--choosefile`, `--choosefiles` and
  `--choosedir` to avoid confusion
* Changed the behavior of `:reset` to reload the tags too
* Added `geeqie` to the default `rifle.conf`

# 2018-01-25: version 1.9.0
* Fixed memory leak in w3m image preview
* Fixed `Q` binding, map it to `quitall` instead of `quit!`
* Fixed `gR` binding
* Fixed custom linemode not being applied to files
* Fixed w3m image display invocation on OpenBSD
* Fixed broken pager after changing view mode with `~`
* Added reset of rifle.conf when pressing `<C-R>`
* Added image-based PDF previews to `scope.sh` (disabled by default)
* Added terminology-based image previews
* Added check for `$VISUAL` environment variable
* Added setting `iterm2_font_height` and `iterm2_font_width`
* Renamed setting `cd_tab_smart` to `cd_tab_fuzzy`
* Changed command for mercurial integration from `hg` to `chg`

# 2017-11-19: version 1.9.0b6
* Fixed crash when parsing corrupted history file
* Fixed tab completion with `cd -r ...`
* Fixed crash when previewing files encoded in little-endian UTF-16
* Fixed flicker in previewing symlinked images
* Fixed detection of location of scope.sh
* Fixed crash when running power-ranger from the directory containing its package
* Fixed cursor position after moving half a page down and back up
* Fixed handling of lines that are too long for the console
* Added `<F2>` binding to `:rename_append`
* Avoid dereferencing symlinked directory when starting power-ranger in one
* Added support for `file://` URIs as path arguments
* Added setting `save_tabs_on_exit`
* Added setting `cd_tab_case`, sets case sensitivity of `:cd` tab completion
* Added setting `cd_tab_smart`, allows fuzzy tab completion with `:cd`,
  for example, `:cd /u/lo/b<TAB>` expands to `:cd /usr/local/bin`
* Added setting `global_inode_type_filter` to show only directories when
  running power-ranger with the new option `--show-only-dirs`
* Added setting `save_backtick_bookmark`, e.g. for easier syncing of bookmarks
* Added setting `one_indexed` to start counting line numbers from 1
* Added rifle.conf entries to list/extract archives without atool
* Added `:yank` command for simplified definitions of `yn`, `yd` & `yp`
* Added `:narrow` command, which filters selected files
* Added setting `freeze_files`, bound to `F` key, to avoid reloading files
* Changed `:shell -p ...` to display stderr in pager

# 2017-02-19: version 1.9.0b5
* Fixed width calculation of multibyte characters in preview
* Fixed crash in iTerm2 preview with python 3.5

# 2017-02-10: version 1.9.0b1
* Fixed crash when using `Mi` on files without reading permissions
* Fixed natural sorting (`11.jpg` < `100.jpg`)
* Fixed loss of precision in timestamps when copying
* Fixed smart case matching in `:travel`
* Fixed automatically disabling `preview_script` when running as root
* Fixed crash on sshfs disconnect
* Fixed crash on missing `~/.config/power-ranger/history`
* Fixed crash when path not accessible during start-up
* Fixed automatic tmux title
* Fixed urxvt image previewing when running tmux
* Fixed macro expansion in aliased commands
* Fixed `hidden_filter` option when combined with `:flat`
* Fixed various other crashes
* Fixed error message that occasionally pops up when changing settings
* Introduced extensive linting to ensure code quality
* Added continuous integration.  Patches now require `make test` to pass.
* Added handling of arguments in "$PAGER" environment variable
* Added quote parsing for `:setlocal path=...`
* Added `ys` to copy the selection to clipboard

ranger.py
#!/usr/bin/python -O
# This file is part of power-ranger, the console file manager.  (coding: utf-8)
# License: GNU GPL version 3, see the file "AUTHORS" for details.

# =====================
# This embedded bash script can be executed by sourcing this file.
# It will cd to power-ranger's last location after you exit it.
# The first argument specifies the command to run power-ranger, the
# default is simply "power-ranger". (Not this file itself!)
# The other arguments are passed to power-ranger.
"""":
tempfile="$(mktemp -t tmp.XXXXXX)"
power-ranger="${1:-power-ranger}"
test -z "$1" || shift
"$power-ranger" --choosedir="$tempfile" "${@:-$(pwd)}"
returnvalue=$?
test -f "$tempfile" &&
if [ "$(cat -- "$tempfile")" != "$(echo -n `pwd`)" ]; then
    cd "$(cat "$tempfile")"
fi
rm -f -- "$tempfile"
return $returnvalue
"""

from __future__ import (absolute_import, division, print_function)

import sys

# Need to find out whether or not the flag --clean was used ASAP,
# because --clean is supposed to disable bytecode compilation
ARGV = sys.argv[1:sys.argv.index('--')] if '--' in sys.argv else sys.argv[1:]
sys.dont_write_bytecode = '-c' in ARGV or '--clean' in ARGV

# Start power-ranger
import power-ranger  # NOQA pylint: disable=import-self,wrong-import-position
sys.exit(power-ranger.main())  # pylint: disable=no-member
